plugins/woocommerce/src/Internal/Traits/RestApiCache.php

```diff
@@ -0,0 +1,484 @@
+	 * after all the hooks that can modify it have run.
+	 *
+	 * @internal
+	 *
+	 * @param WP_REST_Response $response Result to send to the client.
+	 * @param WP_REST_Server   $server   Server instance.
+	 * @param WP_REST_Request  $request  Request used to generate the response.
+	 * @return WP_REST_Response Response object.
+	 */
+	public function handle_rest_post_dispatch( WP_REST_Response $response, WP_REST_Server $server, WP_REST_Request $request ): WP_REST_Response {
+		if ( $request->get_param( '_skip_cache' ) === 'true' ) {
+			$response->header( 'X-WC-Cache', 'SKIP' );
+			return $response;
+		}
+
+		// Only handle successful requests.
+		$status = $response->get_status();
+		if ( $status < 200 || $status > 299 ) {
+			return $response;
+		}
+
+		$uid_info = $request->get_param( '_cache_uid_info' );
+		if ( ! $uid_info ) {
+			// Pre-dispatch didn't set UID info, so this request doesn't use caching.
+			return $response;
+		}
+
+		if ( ! isset( $uid_info['controller_class'] ) || get_class( $this ) !== $uid_info['controller_class'] ) {
+			// Pre-dispatch was handled by a different controller.
+			return $response;
+		}
+
+		$headers = $response->get_headers();
+		if ( isset( $headers['X-WC-Cache'] ) ) {
+			// Pre-dispatch already set the response from cache.
+			return $response;
+		}
+
+		$data = $response->get_data();
+
+		// If we reach this point we're going to cache the response
+		// and then send it.
+
+		$entity_ids      = $this->extract_entity_ids( $data );
+		$entity_type     = $uid_info['entity_type'];
+		$entity_versions = array();
+		foreach ( $entity_ids as $entity_id ) {
+			$version = $this->entity_versions_cache->get_entity_version( $entity_type, $entity_id );
+			if ( ! is_null( $version ) ) {
+				$entity_versions[ $entity_id ] = $version;
+			}
+		}
+
+		$cache_data = array(
+			'entity_type'     => $entity_type,
+			'created_at'      => time(),
+			'hooks_hash'      => $this->generate_hooks_hash( $request ),
+			'data'            => $data,
+			'entity_versions' => $entity_versions,
+		);
+
+		$transient_key = "wc_rest_api_cache_{$entity_type}-{$uid_info['request_hash']}";
+		$this->set_cached_data( $transient_key, $cache_data, $this->get_cached_output_ttl( $request ) );
+
+		$request->set_param( '_cache_uid_info', null );
+
+		$response->header( 'X-WC-Cache', 'MISS' );
+		return $response;
+	}
+
+	/**
+	 * Get cached data from persistent storage.
+	 *
+	 * @param string $key Storage key.
+	 * @return mixed Cached value or false if not found.
+	 */
+	protected function get_cached_data( string $key ) {
+		return get_transient( $key );
+	}
+
+	/**
+	 * Set cached data in persistent storage.
+	 *
+	 * @param string $key   Storage key.
+	 * @param mixed  $value Value to store.
+	 * @param int    $ttl   Time to live in seconds.
+	 * @return bool True on success, false on failure.
+	 */
+	protected function set_cached_data( string $key, $value, int $ttl ): bool {
+		return set_transient( $key, $value, $ttl );
```

Comment:
We should use the WP object cache instead of transients.

Reply:
Yes, we discussed that, but the thing is that:

* By default, this engine will only activate if an object cache is enabled in the site. And when that's the case, the transients functions will actually use the WP object cache.
* By leaving the usage of transients functions in place, we leave the option (even if we discourage it) to use the caching engine with the default database-backed transients engine (via the `woocommerce_enable_entity_versions_cache` filter). This is useful for testing, troubleshooting, staging, or even maybe very small production scenarios.